### PR TITLE
[clang][Sema] Skip the RequiresExprBodyDecls for lambda dependencies

### LINKED ISSUE #56556: clang error when partial specialization with a single template parameter
I define an adapter class Adapter and two template classes A<...>, B<...>, when I partially specialize Adapter for A and B, for some consideration, I use the requires clause instead of traditional way (like this: template \<typename T1\> struct Adapter\<A\<T1>> {...}), for example:
```cpp
// partial specialization for A<...>
template <typename TType> requires true
	&& IsTemplateInstanceOfT<TType, A> // ok if msvc/gcc, failed if clang
struct Adapter<TType> {
	static auto fun (
	) -> void {
		std::cout << "A" << std::endl;
		return;
	}
};
```
IsTemplateInstanceOfT<TType, A> is a concept to check TType is a A(class template)'s instance or not

I used msvc, gcc, and clang to test this code. When I only partially specialize for A, all three can be successfully compiled, but if I perform partial specialization for A and B at the same time, both msvc and gcc can be successful. But clang doesn't allow this code, it tells me there are ambiguous partial specializations of 'Adapter<A<int>>' and 'Adapter<B<int>>'
```
y.cpp:59:2: error: ambiguous partial specializations of 'Adapter<A<int>>'
        Adapter<A<int>>::fun(); // print : A
        ^
y.cpp:34:8: note: partial specialization matches [with TType = A<int>]
struct Adapter<TType> {
       ^
y.cpp:46:8: note: partial specialization matches [with TType = A<int>]
struct Adapter<TType> {
       ^
1 error generated.
```

To circumvent this error, I need to use std::is_base_of to make clang allow my code, like this:
```cpp
// partial specialization for A<...>
template <typename TType> requires true
	&& std::is_base_of_v<A<typename TType::Value>, TType> // ok if msvc/gcc/clang
struct Adapter<TType> {
	static auto fun (
	) -> void {
		std::cout << "A" << std::endl;
		return;
	}
};
```
Is this a bug?

clang version:
```
Ubuntu clang version 15.0.0-++20220627042157+655dc02cb023-1~exp1~20220627042241.261
Target: x86_64-pc-linux-gnu
Thread model: posix
```

Here is the complete test code:
```cpp
#include <iostream>

// declare a value
template <typename It>
inline constexpr auto declare (
) -> It;

// check a type is template's instance or not
template <typename It, template <typename> typename Template>
concept IsTemplateInstanceOfT = requires {
	{ [] <typename T1> (Template<T1> &) {}(declare<It &>()) };
};

// class A
template <typename TValue>
struct A {
	using Value = TValue;
};

// class B
template <typename TValue>
struct B {
	using Value = TValue;
};

// template
template <typename TType>
struct Adapter;

// partial specialization for A<...>
template <typename TType> requires true
	&& IsTemplateInstanceOfT<TType, A> // ok if msvc/gcc, failed if clang
	//&& std::is_base_of_v<A<typename TType::Value>, TType> // ok if msvc/gcc/clang
struct Adapter<TType> {
	static auto fun (
	) -> void {
		std::cout << "A" << std::endl;
		return;
	}
};

// partial specialization for B<...>
template <typename TType> requires true
	&& IsTemplateInstanceOfT<TType, B> // ok if msvc/gcc, failed if clang
	//&& std::is_base_of_v<B<typename TType::Value>, TType> // ok if msvc/gcc/clang
struct Adapter<TType> {
	static auto fun (
	) -> void {
		std::cout << "B" << std::endl;
		return;
	}
};

// test
auto main (
) -> int {
	std::cout << "A<int> is A ? " << IsTemplateInstanceOfT<A<int>, A> << std::endl; // print : A<int> is A ? 1
	std::cout << "B<int> is A ? " << IsTemplateInstanceOfT<B<int>, A> << std::endl; // print : B<int> is A ? 0
	Adapter<A<int>>::fun(); // print : A
	Adapter<B<int>>::fun(); // print : B
	return 0;
}

```

### LINKED ISSUE #82849: Constant Evaluation Assertion Failure with C++ Concepts
If a concept contains a lambda expression, and if the arguments provided to the concept template-id expression are not dependent, `Sema` runs into an assertion failure. 

Consider the following program:
```
template <typename> 
concept C = requires {
  [] {};
};

template <typename>
void b() {
  C<int>;
}
```

The assertion failure is 
```
clang++: ~/llvm-project/clang/lib/AST/ExprConstant.cpp:15739: bool clang::Expr::EvaluateAsConstantExpr(EvalResult &, const ASTContext &, ConstantExprKind) const: Assertion `!isValueDependent() && "Expression evaluator can't be called on a dependent expression."' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: /home/ych/llvm-project/build/bin/clang++ -cc1 -std=c++20 crash.cpp
1.      crash.cpp:8:9: current parser token ';'
2.      crash.cpp:7:10: parsing function body 'b'
3.      crash.cpp:7:10: in compound statement ('{}')
 #0 0x000055643425386d llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /home/ych/llvm-project/llvm/lib/Support/Unix/Signals.inc:723:11
 #1 0x0000556434253d5b PrintStackTraceSignalHandler(void*) /home/ych/llvm-project/llvm/lib/Support/Unix/Signals.inc:798:1
 #2 0x0000556434251dc6 llvm::sys::RunSignalHandlers() /home/ych/llvm-project/llvm/lib/Support/Signals.cpp:105:5
 #3 0x0000556434254515 SignalHandler(int) /home/ych/llvm-project/llvm/lib/Support/Unix/Signals.inc:413:1
 #4 0x00007f2777e54db0 __restore_rt (/lib64/libc.so.6+0x54db0)
 #5 0x00007f2777ea365c __pthread_kill_implementation (/lib64/libc.so.6+0xa365c)
 #6 0x00007f2777e54d06 gsignal (/lib64/libc.so.6+0x54d06)
 #7 0x00007f2777e287f3 abort (/lib64/libc.so.6+0x287f3)
 #8 0x00007f2777e2871b _nl_load_domain.cold (/lib64/libc.so.6+0x2871b)
 #9 0x00007f2777e4dca6 (/lib64/libc.so.6+0x4dca6)
#10 0x000055643abfc801 clang::Expr::EvaluateAsConstantExpr(clang::Expr::EvalResult&, clang::ASTContext const&, clang::Expr::ConstantExprKind) const /home/ych/llvm-project/clang/lib/AST/ExprConstant.cpp:0:3
#11 0x0000556438f1a7a3 clang::ActionResult<clang::Expr*, true> calculateConstraintSatisfaction<calculateConstraintSatisfaction(clang::Sema&, clang::NamedDecl const*, clang::SourceLocation, clang::MultiLevelTemplateArgumentList const&, clang::Expr const*, clang::ConstraintSatisfaction&)::$_0>(clang::Sema&, clang::Expr const*, clang::ConstraintSatisfaction&, calculateConstraintSatisfaction(clang::Sema&, clang::NamedDecl const*, clang::SourceLocation, clang::MultiLevelTemplateArgumentList const&, clang::Expr const*, clang::ConstraintSatisfaction&)::$_0&&) /home/ych/llvm-project/clang/lib/Sema/SemaConcept.cpp:289:71
#12 0x0000556438f1a26c calculateConstraintSatisfaction(clang::Sema&, clang::NamedDecl const*, clang::SourceLocation, clang::MultiLevelTemplateArgumentList const&, clang::Expr const*, clang::ConstraintSatisfaction&) /home/ych/llvm-project/clang/lib/Sema/SemaConcept.cpp:339:10
#13 0x0000556438f13926 CheckConstraintSatisfaction(clang::Sema&, clang::NamedDecl const*, llvm::ArrayRef<clang::Expr const*>, llvm::SmallVectorImpl<clang::Expr*>&, clang::MultiLevelTemplateArgumentList const&, clang::SourceRange, clang::ConstraintSatisfaction&) /home/ych/llvm-project/clang/lib/Sema/SemaConcept.cpp:454:22
#14 0x0000556438f13623 clang::Sema::CheckConstraintSatisfaction(clang::NamedDecl const*, llvm::ArrayRef<clang::Expr const*>, llvm::SmallVectorImpl<clang::Expr*>&, clang::MultiLevelTemplateArgumentList const&, clang::SourceRange, clang::ConstraintSatisfaction&) /home/ych/llvm-project/clang/lib/Sema/SemaConcept.cpp:509:7
#15 0x0000556438f80ac4 clang::Sema::CheckConstraintSatisfaction(clang::NamedDecl const*, llvm::ArrayRef<clang::Expr const*>, clang::MultiLevelTemplateArgumentList const&, clang::SourceRange, clang::ConstraintSatisfaction&) /home/ych/llvm-project/clang/include/clang/Sema/Sema.h:7680:12
#16 0x0000556439dbc13b clang::Sema::CheckConceptTemplateId(clang::CXXScopeSpec const&, clang::SourceLocation, clang::DeclarationNameInfo const&, clang::NamedDecl*, clang::ConceptDecl*, clang::TemplateArgumentListInfo const*) /home/ych/llvm-project/clang/lib/Sema/SemaTemplate.cpp:5024:7
#17 0x0000556439dbc5b8 clang::Sema::BuildTemplateIdExpr(clang::CXXScopeSpec const&, clang::SourceLocation, clang::LookupResult&, bool, clang::TemplateArgumentListInfo const*) /home/ych/llvm-project/clang/lib/Sema/SemaTemplate.cpp:5077:12
#18 0x000055643950be62 clang::Sema::ActOnIdExpression(clang::Scope*, clang::CXXScopeSpec&, clang::SourceLocation, clang::UnqualifiedId&, bool, bool, clang::CorrectionCandidateCallback*, bool, clang::Token*) /home/ych/llvm-project/clang/lib/Sema/SemaExpr.cpp:2927:12
#19 0x0000556438accc2b clang::Parser::tryParseCXXIdExpression(clang::CXXScopeSpec&, bool, clang::Token&) /home/ych/llvm-project/clang/lib/Parse/ParseExprCXX.cpp:641:17
#20 0x0000556438ace58a clang::Parser::ParseCXXIdExpression(bool) /home/ych/llvm-project/clang/lib/Parse/ParseExprCXX.cpp:734:7
#21 0x0000556438ab8b50 clang::Parser::ParseCastExpression(clang::Parser::CastParseKind, bool, bool&, clang::Parser::TypeCastState, bool, bool*) /home/ych/llvm-project/clang/lib/Parse/ParseExpr.cpp:1687:11
#22 0x0000556438ab07c9 clang::Parser::ParseCastExpression(clang::Parser::CastParseKind, bool, clang::Parser::TypeCastState, bool, bool*) /home/ych/llvm-project/clang/lib/Parse/ParseExpr.cpp:693:20
#23 0x0000556438aaecb8 clang::Parser::ParseAssignmentExpression(clang::Parser::TypeCastState) /home/ych/llvm-project/clang/lib/Parse/ParseExpr.cpp:176:20
#24 0x0000556438aaeb6f clang::Parser::ParseExpression(clang::Parser::TypeCastState) /home/ych/llvm-project/clang/lib/Parse/ParseExpr.cpp:127:18
#25 0x0000556438b472c8 clang::Parser::ParseExprStatement(clang::Parser::ParsedStmtContext) /home/ych/llvm-project/clang/lib/Parse/ParseStmt.cpp:539:19
#26 0x0000556438b45678 clang::Parser::ParseStatementOrDeclarationAfterAttributes(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*, clang::ParsedAttributes&, clang::ParsedAttributes&) /home/ych/llvm-project/clang/lib/Parse/ParseStmt.cpp:277:14
#27 0x0000556438b44d4b clang::Parser::ParseStatementOrDeclaration(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*) /home/ych/llvm-project/clang/lib/Parse/ParseStmt.cpp:118:20
#28 0x0000556438b4dec2 clang::Parser::ParseCompoundStatementBody(bool) /home/ych/llvm-project/clang/lib/Parse/ParseStmt.cpp:1236:11
#29 0x0000556438b4f5b1 clang::Parser::ParseFunctionStatementBody(clang::Decl*, clang::Parser::ParseScope&) /home/ych/llvm-project/clang/lib/Parse/ParseStmt.cpp:2514:21
#30 0x0000556438a77d34 clang::Parser::ParseFunctionDefinition(clang::ParsingDeclarator&, clang::Parser::ParsedTemplateInfo const&, clang::Parser::LateParsedAttrList*) /home/ych/llvm-project/clang/lib/Parse/Parser.cpp:1517:3
#31 0x0000556438b1e41d clang::Parser::ParseDeclGroup(clang::ParsingDeclSpec&, clang::DeclaratorContext, clang::ParsedAttributes&, clang::Parser::ParsedTemplateInfo&, clang::SourceLocation*, clang::Parser::ForRangeInit*) /home/ych/llvm-project/clang/lib/Parse/ParseDecl.cpp:2320:21
#32 0x0000556438b66c33 clang::Parser::ParseDeclarationAfterTemplate(clang::DeclaratorContext, clang::Parser::ParsedTemplateInfo&, clang::ParsingDeclRAIIObject&, clang::SourceLocation&, clang::ParsedAttributes&, clang::AccessSpecifier) /home/ych/llvm-project/clang/lib/Parse/ParseTemplate.cpp:258:10
#33 0x0000556438b65da8 clang::Parser::ParseTemplateDeclarationOrSpecialization(clang::DeclaratorContext, clang::SourceLocation&, clang::ParsedAttributes&, clang::AccessSpecifier) /home/ych/llvm-project/clang/lib/Parse/ParseTemplate.cpp:174:10
#34 0x0000556438b657a0 clang::Parser::ParseDeclarationStartingWithTemplate(clang::DeclaratorContext, clang::SourceLocation&, clang::ParsedAttributes&) /home/ych/llvm-project/clang/lib/Parse/ParseTemplate.cpp:50:10
#35 0x0000556438b1ccbe clang::Parser::ParseDeclaration(clang::DeclaratorContext, clang::SourceLocation&, clang::ParsedAttributes&, clang::ParsedAttributes&, clang::SourceLocation*) /home/ych/llvm-project/clang/lib/Parse/ParseDecl.cpp:1919:12
#36 0x0000556438a7535e clang::Parser::ParseExternalDeclaration(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec*) /home/ych/llvm-project/clang/lib/Parse/Parser.cpp:986:14
#37 0x0000556438a7372c clang::Parser::ParseTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&, clang::Sema::ModuleImportState&) /home/ych/llvm-project/clang/lib/Parse/Parser.cpp:760:12
#38 0x0000556438a6e4fa clang::ParseAST(clang::Sema&, bool, bool) /home/ych/llvm-project/clang/lib/Parse/ParseAST.cpp:163:16
#39 0x000055643564e06b clang::ASTFrontendAction::ExecuteAction() /home/ych/llvm-project/clang/lib/Frontend/FrontendAction.cpp:1185:1
#40 0x000055643564da8c clang::FrontendAction::Execute() /home/ych/llvm-project/clang/lib/Frontend/FrontendAction.cpp:1073:7
#41 0x000055643556daa7 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) /home/ych/llvm-project/clang/lib/Frontend/CompilerInstance.cpp:1057:23
#42 0x000055643581eac1 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) /home/ych/llvm-project/clang/lib/FrontendTool/ExecuteCompilerInvocation.cpp:272:8
#43 0x0000556432446a7b cc1_main(llvm::ArrayRef<char const*>, char const*, void*) /home/ych/llvm-project/clang/tools/driver/cc1_main.cpp:232:13
#44 0x0000556432437002 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) /home/ych/llvm-project/clang/tools/driver/driver.cpp:365:5
#45 0x0000556432435d5b clang_main(int, char**, llvm::ToolContext const&) /home/ych/llvm-project/clang/tools/driver/driver.cpp:406:5
#46 0x000055643246b3b5 main /home/ych/llvm-project/build/tools/clang/tools/driver/clang-driver.cpp:17:3
#47 0x00007f2777e3feb0 __libc_start_call_main (/lib64/libc.so.6+0x3feb0)
#48 0x00007f2777e3ff60 __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x3ff60)
#49 0x00005564324357e5 _start (/home/ych/llvm-project/build/bin/clang+++0x57017e5)
fish: Job 1, '~/llvm-project/build/bin/clang+…' terminated by signal SIGABRT (Abort)
```

`ExprConstant` complains that this lambda after instantiation is still value dependent. `getTypeDeclType` on the instantiated lambda shows that the type for the lambda `CXXRecordDecl` is dependent. The reason being the parent `DeclContext` was dependent, which is also true in this case as we aren't instantiating function template `b` here. 

It's worth noting that with an assertion-disabled `clang++`, this program compiles successfully and, as far as I can tell, correctly.

I can think of a few possible fixes:
- Why does a `template-id` expression here causes the constraint to be checked if the function template `b` is not being instantiated? Can the fix just be that defer checking the constraint to performing pending instantiations? In `Sema::CheckConceptTemplateId`, we eagerly calculate the constraint satisfaction if the arguments to the concept template aren't dependent.  
- Why does `DeclContext` look at the parent context to tell if it's instantiated? This causes any AST nodes created with template instantiation we do during concept template constraints checking to be considered dependent. 
- Defer checking value dependency at `ExprConstant` evaluation. Since in our case, no actual dependent values can be reached, which explains why the assertion-free build successfully compiles this code. 

Even though a Release build is not affected and won't block any production users (with doubt), I think this problem is still worth solving. I would like to consult upstream which approach I should take.

### PR DESCRIPTION
**Pull request by zyn0217**
The dependency of a lambda inside of a `RequiresExprBodyDecl` was previously affected by its parent, e.g., `ClassTemplateSpecializationDecl`. This made the lambda always dependent regardless of the template arguments we had, which caused some crashes on the constraint evaluation later.

This fixes https://github.com/llvm/llvm-project/issues/56556, fixes https://github.com/llvm/llvm-project/issues/82849 and a case demonstrated by https://github.com/llvm/llvm-project/issues/49570#issuecomment-1664966972.

### CODE REVIEW

File: clang/lib/Sema/TreeTransform.h [hunk 1]
```diff
@@ -13649,10 +13649,29 @@ TreeTransform<Derived>::TransformLambdaExpr(LambdaExpr *E) {
...
+  // template <C F>
+  // struct Widget<F> {};
+  //
+  // While we are here in substitution for Widget<F>, the parent of DC would be
+  // the template specialization itself. Thus, the lambda expression
+  // will be deemed as dependent even if we have non-dependent template
+  // arguments.
+  // (A ClassTemplateSpecializationDecl is always a dependent context.)
+  if (DC->getDeclKind() == Decl::Kind::RequiresExprBody)
+    DC = DC->getParent();
```
Comment (cor3ntin) [inline clang/lib/Sema/TreeTransform.h:13671]: Do we get into the same issue if there are multiple nested `requires`?

File: clang/lib/Sema/TreeTransform.h [hunk 2]
```diff
@@ -13649,10 +13649,29 @@ TreeTransform<Derived>::TransformLambdaExpr(LambdaExpr *E) {
...
+  // template <typename>
+  // concept C = requires { [] {}; };
+  //
+  // template <class F>
+  // struct Widget;
+  //
+  // template <C F>
+  // struct Widget<F> {};
+  //
+  // While we are here in substitution for Widget<F>, the parent of DC would be
```
Comment (cor3ntin) [inline clang/lib/Sema/TreeTransform.h:13665]:
```suggestion
  // While we are substituting Widget<F>, the parent of DC would be
```

File: clang/lib/Sema/TreeTransform.h [hunk 3]
```diff
@@ -13649,10 +13649,29 @@ TreeTransform<Derived>::TransformLambdaExpr(LambdaExpr *E) {
...
+  // the template specialization itself. Thus, the lambda expression
+  // will be deemed as dependent even if we have non-dependent template
+  // arguments.
```
Comment (cor3ntin) [inline clang/lib/Sema/TreeTransform.h:13668]:
```suggestion
  // will be deemed as dependent even if there is are no dependent template
  // arguments.
```

File: clang/lib/Sema/TreeTransform.h (hunk 1 above)
Comment (zyn0217) [reply clang/lib/Sema/TreeTransform.h]: Oops, this should be a `while` loop, I think

**Comment by zyn0217** [comment]
@cor3ntin Thanks! I have updated the test to reflect the nested `requires` cases. PTAL?

**Review by cor3ntin — APPROVED** [review]
LGTM

**Review by yuxuanchen1997 — APPROVED** [review]
Thank you so much! This looks good.

**Comment by alexfh** [comment]
Hi folks, we've started seeing "implicit instantiation of undefined template" compilation errors after this commit. In all cases code used to compile fine before this commit. I haven't spotted any particular issues in the code (though the examples are looked at are rather convoluted, thus, I'm not 100% sure). I'll try to prepare a self-sufficient test case.

Was this sort of a behavior change expected?

**Comment by zyn0217** [comment]
> Was this sort of a behavior change expected?

I don't think so; a reproducer would be really appreciated.

**Comment by alexfh** [comment]
I'm reducing the test case. In the meantime, should this be reverted?

**Comment by alexfh** [comment]
A bit more compact one: https://gcc.godbolt.org/z/4rzYPKaze

**Comment by jyknight** [comment]
Can reduce further to:
```
template <typename T>
concept h = requires(T i) { [] {}(i); };

template <typename T> struct k;
template <h m> struct k<m> {
  struct n;
};
using o = k<int>::n;
```

But, is `requires(T i) { [] {}(i); };` actually valid? I think that _should_ fail the requirement, since you cannot call that lambda with an argument? Maybe that's just an artifact of the reduction?

**Comment by dwblaikie** [comment]
FWIW, @jyknight's example fails to compile with GCC - succeeds with clang 18 release but assert-fails with clang 18 +Asserts build. (not sure about the original/full internal reproduction - we have a way to run compile with assertions enaled, but I'm not sure I'm holding it right... )

**Comment by zyn0217** [comment]
> I have a reduced test case, which is accepted by clang 18, but fails with this patch: https://gcc.godbolt.org/z/zMbKvsf7K
> A bit more compact one: https://gcc.godbolt.org/z/4rzYPKaze

Your case is rejected by all mainstream compilers as of now: https://gcc.godbolt.org/z/nrMf3zvfY
Again, it was an accept-invalid before because of the constant evaluation on dependent expressions, which **is** an assertion failure if you compile it with debug mode. https://gcc.godbolt.org/z/eascd7j1G

To clarify why the requirement should be evaluated to false:

> The enclosing _requires-expression_ will evaluate to false if substitution of template arguments into the expression fails.

https://eel.is/c++draft/expr.prim.req#simple-1.sentence-2

**Comment by alexfh** [comment]
I guess the reduction could have dropped some important parts of this. Let me try the original code with assertions-enabled clang build...

**Comment by alexfh** [comment]
It turns out debug build of clang before this patch generated an assertion on the original code as well:
```
assert.h assertion failed at llvm-project/clang/lib/AST/ExprConstant.cpp:15739 in bool clang::Expr::EvaluateAsConstantExpr(EvalResult &, const ASTContext &, ConstantExprKind) const: !isValueDependent() && "Expression evaluator can't be called on a dependent expression."
...
    @     0x5643bcdac9e4  __assert_fail
    @     0x5643b9734df7  clang::Expr::EvaluateAsConstantExpr()
    @     0x5643b89f7fea  calculateConstraintSatisfaction<>()
    @     0x5643b89f1bcf  CheckConstraintSatisfaction()
    @     0x5643b89f193e  clang::Sema::CheckConstraintSatisfaction()
    @     0x5643b90d9f23  clang::Sema::CheckConceptTemplateId()
    @     0x5643b9259656  clang::TreeTransform<>::TransformConceptSpecializationExpr()
    @     0x5643b925ce00  clang::TreeTransform<>::TransformCXXFoldExpr()
    @     0x5643b924b6e0  clang::Sema::SubstConstraintExpr()
    @     0x5643b89f9097  calculateConstraintSatisfaction()::$_0::operator()()
    @     0x5643b89f7f50  calculateConstraintSatisfaction<>()
    @     0x5643b89f1bcf  CheckConstraintSatisfaction()
    @     0x5643b89f193e  clang::Sema::CheckConstraintSatisfaction()
    @     0x5643b9237c6d  FinishTemplateArgumentDeduction<>()
    @     0x5643b9236cd2  llvm::function_ref<>::callback_fn<>()
    @     0x5643b888fa2f  clang::Sema::runWithSufficientStackSpace()
    @     0x5643b91c2e17  clang::Sema::DeduceTemplateArguments()
    @     0x5643b9248e19  clang::Sema::InstantiateClassTemplateSpecialization()
    @     0x5643b9334e73  llvm::function_ref<>::callback_fn<>()
    @     0x5643b888fa2f  clang::Sema::runWithSufficientStackSpace()
    @     0x5643b931fdc4  clang::Sema::RequireCompleteTypeImpl()
    @     0x5643b931f535  clang::Sema::RequireCompleteType()
    @     0x5643b891bad7  clang::Sema::RequireCompleteDeclContext()
    @     0x5643b90eda4d  clang::Sema::CheckTypenameType()
    @     0x5643b926edd2  clang::TreeTransform<>::TransformDependentNameType()
    @     0x5643b9243d1c  clang::TreeTransform<>::TransformType()
    @     0x5643b92436ff  clang::TreeTransform<>::TransformType()
    @     0x5643b9244293  clang::Sema::SubstType()
    @     0x5643b90d0e8a  clang::Sema::CheckTemplateIdType()
    @     0x5643b90d5015  clang::Sema::ActOnTemplateIdType()
    @     0x5643b8674084  clang::Parser::AnnotateTemplateIdTokenAsType()
    @     0x5643b86536cf  clang::Parser::ParseDeclarationSpecifiers()
    @     0x5643b864fffa  clang::Parser::ParseSpecifierQualifierList()
    @     0x5643b863da31  clang::Parser::ParseTypeName()
    @     0x5643b86262bd  clang::Parser::ParseAliasDeclarationAfterDeclarator()
    @     0x5643b8624ea0  clang::Parser::ParseUsingDeclaration()
    @     0x5643b8623e70  clang::Parser::ParseUsingDirectiveOrDeclaration()
    @     0x5643b864b6ef  clang::Parser::ParseDeclaration()
    @     0x5643b868a9a4  clang::Parser::ParseStatementOrDeclarationAfterAttributes()
    @     0x5643b8689d78  clang::Parser::ParseStatementOrDeclaration()
    @     0x5643b8693bc0  clang::Parser::ParseCompoundStatementBody()
    @     0x5643b8694bce  clang::Parser::ParseFunctionStatementBody()
    @     0x5643b85da8ad  clang::Parser::ParseFunctionDefinition()
    @     0x5643b864df89  clang::Parser::ParseDeclGroup()
    @     0x5643b85d8ec9  clang::Parser::ParseDeclOrFunctionDefInternal()
    @     0x5643b85d868e  clang::Parser::ParseDeclarationOrFunctionDefinition()
    @     0x5643b85d7434  clang::Parser::ParseExternalDeclaration()
    @     0x5643b8622693  clang::Parser::ParseInnerNamespace()
    @     0x5643b8621893  clang::Parser::ParseNamespace()
    @     0x5643b864b7de  clang::Parser::ParseDeclaration()
    @     0x5643b85d6fb6  clang::Parser::ParseExternalDeclaration()
    @     0x5643b85d516b  clang::Parser::ParseTopLevelDecl()
    @     0x5643b85ceebe  clang::ParseAST()
    @     0x5643b831a4c3  clang::FrontendAction::Execute()
    @     0x5643b8294efd  clang::CompilerInstance::ExecuteAction()
    @     0x5643b717f84e  clang::ExecuteCompilerInvocation()
    @     0x5643b71737c6  cc1_main()
    @     0x5643b7171066  ExecuteCC1Tool()
    @     0x5643b843fe3e  llvm::function_ref<>::callback_fn<>()
    @     0x5643bcc44415  llvm::CrashRecoveryContext::RunSafely()
    @     0x5643b843f5bb  clang::driver::CC1Command::Execute()
    @     0x5643b8400494  clang::driver::Compilation::ExecuteCommand()
    @     0x5643b84007af  clang::driver::Compilation::ExecuteJobs()
    @     0x5643b841ea80  clang::driver::Driver::ExecuteCompilation()
    @     0x5643b717061b  clang_main()
    @     0x5643b716d834  main
```

I'll try to figure out what's wrong with the original code.

**Comment by alexfh** [comment]
The internal code was fixed. Thanks everyone for helping to figure this out!

**Pull request merged**